### ISSUE TITLE
Iterative version of function articulation

### ISSUE THREAD

**Issue by simonschoelly** [JuliaGraphs/LightGraphs.jl#1106]
The function `articulation` is recursive at the moment which leads to stack overflows on large graphs.

For example this code gives a stack overflow on my computer.
```
julia> articulation(PathGraph(5 * 10^4))
ERROR: StackOverflowError:
```

**Comment by abhinavmehndiratta** [comment]
A bit more optimized code -
```julia
function iterative_articulation(g::AbstractGraph)
    s = Array{Int}(undef,0)
    visited = falses(nv(g))
    parent = zeros(Int,nv(g))
    low = zeros(Int,nv(g))
    disc = zeros(Int,nv(g))
    articulation_points = falses(nv(g))

    for k in vertices(g)
        if visited[k]
            continue
        end
        dfs = Array{Int}(undef,0)
        push!(s,k)
        while !isempty(s)
            u = s[length(s)]
            pop!(s)
            if !visited[u]
                push!(dfs,u)
                low[u] = length(dfs)
                disc[u] = length(dfs)
                visited[u] = true
            end
            for v in outneighbors(g,u)
                if !visited[v]
                    push!(s,v)
                    parent[v] = u
                end
            end
        end
        j = length(dfs)
        k_nchild = 0
        while j > 1
            u = dfs[j]
            for v in outneighbors(g,u)
                if v == parent[u]
                    continue
                end
                low[u] = min(low[u],low[v])
            end
            p = parent[u]
            low[p] = min(low[p],low[u])
            if p!=k && low[u] >= disc[p]
                articulation_points[p] = true
            end
            if p == k
                k_nchild += 1
            end
            j -= 1
        end
        if k_nchild > 1
            articulation_points[k] = true
        end
    end
    return findall(articulation_points)
end
```

**Comment by abhinavmehndiratta** [comment]
@simonschoelly I think this should be fixed by #1111 ?

**Comment by simonschoelly** [comment]
You are right. Thanks.
